### DISCUSSION TITLE
epaper29 not working

### DISCUSSION THREAD

**Discussion by gck303** [arendst/Tasmota#11235] [Q&A] [answered]
I have been going for a couple of days trying to get an SPI based epaper29 display to work. It will not do anything. I have a working  I2C LCD20x4 display, so I think I hoping I would be able to get the epaper to work okay too. 

I have compiled from the source using the mandatory three options (USE_SPI, USE_DISPLAY and USE_DISPLAY_EPAPER_29) . Nothing happens. I have tried using the pre-compiled -display bin, same result. Incidentally, I cannot tell if the pre-compiled bin file works with an epaper display. 

It is connected as:
display--->esp12
VCC--->3.3v
GND--->gnd
SDI---gpio13 (sspimosi)
SCLK --- gpio14 (sspi sclk)
CS--->gpio 12 (sspi cs)
d/c --->gnd
reset--->3.3v
busy ---> [not connected]

I have put an oscilloscope in the esp12's device pins and there seems to be no signal on any of the three wires. I have the RESET on the display connect to 3.3v, the d/c to GND. I am not using any pullup/down resistors. 

This suggests that the SPI is not being enabled and is not working. 

Can anyone make any helpful suggestions?

Thanks, George

![image](https://user-images.githubusercontent.com/30334661/110174177-a5622980-7df7-11eb-85eb-031d40035ec2.png)

**Comment by sfromis** [comment, marked as answer]
`USE_DISPLAY_EPAPER_29` [is included](https://github.com/arendst/Tasmota/blob/development/BUILDS.md) in precompiled tasmota-sensors.bin.gz

No experience with that display. You should probably configure gpio12 as `EPaper 29 CS` instead of `SSPI CS`. This config is to let the driver know that it is that specific display type you've attached.

**Reply by gck303** [reply]
Thanks. I have changed the setting for the CS and am using the pre-compiled -display build (9.3.1). Still nothing. 

I have traced the pins when sending a command **displaytext [izd]ggg[d]**

I see that:
- the clock pin sends a signal shortly after the command is issued (yellow in the trace)
- the CS pin never goes low. Should this need a pullup/down resistor?
- no signal is send on the MOSI pin. This remains at high for the entire duration of the clock signal and after it (green on the trace)

Could it be that the epaper does not work and I need to raise a formal issue with Tasmota?

Is my display defective? Given that no data is being sent on the MOSI pin, I don't think we can say it is a defective display. 

![image](https://user-images.githubusercontent.com/30334661/110203792-fd387900-7e67-11eb-8ab2-32518945c74d.png)

**Reply by gemu2015** [reply]
you should get signals on ALL 3 involved lines. are you sure about these findings? that is not plausible
at all. no pullup or pulldown needed.
i have 2 displays in use, one on an esp8266 and one on an esp32. 
btw. the driver supports only the waveshare brand, jumpered to 3 wire SPI

**Reply by gck303** [reply]
I have changed my ESP12 device and I now am getting on all three pins. I has miss- soldered on the original ESP12, though even when correctly soldered I am still not getting a CS signal. I started again with a different ESP12 and we are now getting a bit of progress. 

I bought this, which is described as _'2.9 Inch Epaper E-paper E-Ink EInk Display Screen SPI Module For ESP32 Arduino'_
https://www.ebay.co.uk/itm/2-9-Inch-Epaper-E-paper-E-Ink-EInk-Display-Screen-SPI-Module-For-ESP32-Arduino/254706100623?var=554481416036

Will this not work with the tasmota software?

Even though the SPI interface is a standard, and there are a number of 2.9 epaper displays will the tasmota _only_ work the waveshare ones?

Is this the correct product:
https://www.waveshare.com/product/displays/e-paper/epaper-2/2.9inch-e-paper-module.htm

**Reply by gemu2015** [reply]
your product also has 3 wire SPI mode. check the (mechanic) switch at the bottom.
however every brand has its own controller setup. from the description of your product it seems to be a hacked full update display, not partial update like waveshare.
so i doubt that it works, but give it a try (with 3 wire mode)
the link above to waveshare is the supported display with partial update, black and white.
i clarified the display doku about waveshare epaper displays.

**Comment by TrDA-hab** [comment]
For your information only!
![001](https://user-images.githubusercontent.com/56403720/110213083-95a61c00-7eaf-11eb-90e7-017242c554dc.jpg)
![photo_2021-03-06_19-04-57](https://user-images.githubusercontent.com/56403720/110213090-9939a300-7eaf-11eb-8a7c-1971ef34009a.jpg)

**Comment by Jason2866** [comment]
Documentation is updated. Only specific Waveshare E-Paper Displays are supported

**Comment by hardlog** [comment]
Hi,

I have the same problem.
Have you already been able to solve it?

I have an ESPS32-WROOM (also with a wmos D1 mini) with 2.9 waveshare epaper.
Same wiring.

Which firmware do you use (tasmota32.bin, tasmota32-display.bin, self-compiled?).

Do I have to take SPI MOSI or SSPI MOSI?

greeting
Florian

**Comment by gck303** [comment]
Okay. We are getting somewhere. I have acquired a OLED (ILI941) display which uses SPI. I have got this to work, so I have some experience .  But, I still do not have a working epaper display. 

I have obtained a waveshare 29 display and changed the resistor on the rear to 3-wire. ) (https://www.waveshare.com/product/displays/e-paper/epaper-2/2.9inch-e-paper-module.htm)

It is configured to work with an ESP12 module:
VCC--->3.3v
GND--->gnd
DIN---gpio13 (sspimosi) - yellow in the trace
SCLK --- gpio14 (sspi sclk) - green in the trace
CS--->gpio 15 (epaper29 CS)
d/c --->gnd
reset--->3.3v
busy ---> [not connected]

I am getting a signal on the CLK and MOSI pins. The CS is triggering.  But, the display is not responding. 

![image](https://user-images.githubusercontent.com/30334661/111920531-6edb0e80-8a87-11eb-9451-023a318ebf30.png)

![image](https://user-images.githubusercontent.com/30334661/111920552-874b2900-8a87-11eb-93cf-76f04d652214.png)

**Reply by gemu2015** [reply]
from your screenshot you selected SPI but you MUST use SSPI.

**Reply by gck303** [reply]
Why? Is this described on any of the documentation that the SSPI settings should be used?

It works using SPI with an SPI OLED display. The only mention I can see in the documentation is for SSPI to be used with an ESP32, and I am using an ESP12. 

Also, the https://tasmota.github.io/docs/Components/ page says the SPI should be used for the epaper display:

66 | 768 | SPI CS | SPI Interface (ePaper Display)
67 | 800 | SPI DC | SPI Interface (ePaper Display)

Incidentally, I had already tried that and without success. But, I would much rather know what I SHOULD be doing rather than taking a scattergun approach with the hope that I eventually stumble on the correct combination that works.

**Reply by Jason2866** [reply]
Mentioned here https://tasmota.github.io/docs/Displays/#hardware-connections
Since the Docu is Open Source too, enhancements to clarify are welcome

**Reply by gck303** [reply]
Where in that link?

Are you saying that three wire SPI requires SPI; and 4 wire SPI requires SSPI?

I am very happy to update the documentation when I have figured it out and got it working. But... at this point I cannot get it to work.

**Reply by Jason2866** [reply]
```
E-Paper displays are connected via software 3-wire SPI (CS, SCLK, MOSI).
DC should be connected to GND , Reset to 3.3 V and busy may be left unconnected.
The jumper on the circuit board of the display must be set to 3-wire SPI
```

**Reply by gck303** [reply]
This is all done:
- jumper on the board re-soldered to 3 wire
- connected per the pins you have mentioned
- displaymodel 5
- displaymode 1

Nothing is displayed. This happens when the pins are set to either the SPI or SSPI. I am still not sure of the difference between these and when I should use one or the other. 

Nothing happens when I also:
- displaymode 0
- displaytext [d]ggg

**Comment by gemu2015** [comment]
the driver itself is using software SPI on ESP8266 and ESP32.
if you specify SPI it might happen in conjunction with other drivers that hardwire SPI would be enabled and this would prevent the driver from doing software SPI. so better always use SSPI on SSPI drivers. i will further clarify this in the docs.

what is shown on the console wenn restarting ?
what is shown on cmd display ?

since you see signals on all 3 lines i assume the driver is actually working.

however there should only be activity on a display cmd.

double check all lines and check if the signals arrive at the board pins.

these cmd should display hallo
displaytext[zD0]hallo
displaytext [d]

i have 2 working displays of this kind and no issues.

**Reply by gck303** [reply]
Okay, Thanks. I have tested all this, and there are no errors or messages on restart, and the display is correctly configured. Certainly those messages are being sent. 

I will try to recompile from and see if that helps later. 

I am wondering if it is a bug, or compilation error. Any chance you have a copy of your compiled .bin file?

What version are you using?

Now, when I was running a trace I did seem to be getting dodgy signals on the MOSI pin. Paybe this could be the issue?

![image](https://user-images.githubusercontent.com/30334661/112016957-13f7f480-8b25-11eb-8847-78000a1e0679.png)

**Comment by gemu2015** [comment]
what lines are in your display ?
it seems to be cs and clk not MOSI
can you display all 3 lines at once ?

i compiled from the last dev version, since there were no changes to this driver since some month i don't think there is a bug in the sources.

**Reply by gck303** [reply]
Any change you could send me a copy of your compiled file, please? I would like to eliminate this as the source of the problem. 

See below for another couple of traces. I can only trace two signals as I only have a two channel 'scope. 

In this first trace: yellow = CS; green = MOSI
![image](https://user-images.githubusercontent.com/30334661/112029308-047ea880-8b31-11eb-9b43-0024892e2367.png)

In the this second trace: yellow = CS; green = CLK
![image](https://user-images.githubusercontent.com/30334661/112029258-f3359c00-8b30-11eb-8e87-cd78abebecd5.png)

I have recompiled again. Still same issue.

**Reply by gemu2015** [reply]
the signals look good!
does your cable have a different color coding ?
my colors are blue (mosi), yellow(sclk) and orange (cs).
and black for GND, red for VCC

did you measure the signals directly at the circuit board ?

to be sure that no cable is defective

**Reply by gck303** [reply]
Okay. I have used a entirely different LoLin ESP8266 development board, re-complied from source and taken 'scope probes at the terminals of the board. The display is connected to the board. I have tried a second display. Both epaper boards have been resoldered into 3 wire mode. 

Same results. Nothing. 

I think that the tasmota is not sending the correct signals to the board and can be seen with the 'scope on the display board. They are all there and similar to what I have described above. The other pins are +v/-ve, where appropriate. 

it seems to me that there is a problem with the tasmota code. 

Where next?

Are these commands enough to force a display on the epaper?
displaytext[zD0]hallo
displaytext [d]

**Reply by joba-1** [reply]
I know a non-working e-paper can be frustrating, but still, Tasmota works, if

1) compiled with necessary parameters
2) configured correctly
3) used with a supported display
4) wired correctly
5) display is not defect

You probably have 1, 2 and 4. We (well, not me actually) could verify 1 and 2 if you provide info that is requested for issues. 3 to 5 you have to verify yourself. I suggest you try the GxEPD2_HelloWorld example from the repo https://github.com/ZinggJM/GxEPD2
It supports 6 different 2.9“ displays!
Try them out by following the instructions in the .ino-file. That is how I found out which 2.13“ display variant I have :)

Good luck!

**Reply by rapi3** [reply]
On my LCD display after writing text with `displaytext Hello` display it is turning off until new command `displaytext [O]` maybe you have the same problem ?

**Reply by gck303** [reply]
I have tried getting the waveshare display working with that library, and I have been told it is not supported! 
https://forum.arduino.cc/index.php?topic=733811.0

It does seem that waveshare do not manufacture the screens, but get them from somewhere else?

Maybe Tasmota should be supporting the 'good-display' ones instead, that are the ones supported by the GxEPD2 library?

What library does the Tasmota software use?


> I know a non-working e-paper can be frustrating, but still, Tasmota works, if
> 
> 1. compiled with necessary parameters
> 2. configured correctly
> 3. used with a supported display
> 4. wired correctly
> 5. display is not defect
> 
> You probably have 1, 2 and 4. We (well, not me actually) could verify 1 and 2 if you provide info that is requested for issues. 3 to 5 you have to verify yourself. I suggest you try the GxEPD2_HelloWorld example from the repo https://github.com/ZinggJM/GxEPD2
> It supports 6 different 2.9“ displays!
> Try them out by following the instructions in the .ino-file. That is how I found out which 2.13“ display variant I have :)
> 
> Good luck!

**Reply by gemu2015** [reply]
you should try with the original waveshare EPD repo.
https://github.com/waveshare/e-Paper

that is where i got the original sources from.

and there has been a change to V3 recently!

if you got it to work with this driver i will implement the modifications into tasmota

**Reply by gemu2015** [reply]
a first look into V3 driver show massive differences. they may have changed the controller ic.
in fact the driver is much smaller now.

**Reply by gemu2015** [reply]
a closer look at v3 it is a 3 color display, these can not be supported easily by tasmota. (may be by ignoring one color)

**Reply by gemu2015** [reply]
the driver epd2in9 is the one i ported for Tasmota. (this was in 2018)
then there was a change in the reset circuit of the board in 2020 (reset pulled low when not connected)
but the same driver still worked. 
my second display has the marking v2 and also works with the original driver.

you may try the driver in epd2in9_V2 which is a slightly modified version from epd2in9

**Reply by gck303** [reply]
Just a final update. I have paused this investigation. I will may come back to in coming months. 

Meanwhile I am using a ILI9341 display and gesture controller. It will do what I need, plus I can get it to work without issue.

**Reply by marmox** [reply]
Undoubtedly an update is needed to support the latest version of Waveshare displays.
If I control the display with Arduino (with 4 wire SPI jumper), loading the code suitable for the latest version of the displays, it works. Tasmota (with 3 wire SPI jumper) not!
